inspirehep/modules/workflows/tasks/actions.py

```diff
@@ -1011,16 +1005,26 @@ def normalize_affiliations(obj, eng):
                 continue
             matched_author_affiliations = _match_lit_author_affiliation(raw_aff)
             if matched_author_affiliations:
-                assigned_affiliations = _assign_matched_affiliations_to_author(
-                    author_affiliations, matched_author_affiliations
-                )
-                matched_affiliations[raw_aff] = assigned_affiliations
+                _assign_institution_reference_to_incomplete_affiliations(matched_author_affiliations)
+                matched_affiliations[raw_aff] = matched_author_affiliations
+                author_affiliations.extend(matched_author_affiliations)
         if author_affiliations:
             author["affiliations"] = author_affiliations
             LOGGER.info(
-                "Normalized affiliations for author",
-                author=author["full_name"],
-                raw_affiliations=raw_affs,
-                assigned_affiliations=author_affiliations,
+                "(wf: {0}) Normalized affiliations for author {1}. Raw affiliations: {2}. Assigned affiliations: {3}".format(
+                    obj.id,
+                    author["full_name"],
+                    " ".join(raw_affs),
+                    author_affiliations
+                )
             )
     return obj
+
+
+def link_institutions_with_affiliations(obj, eng):
+    for author in obj.data.get("authors", []):
```

Comment:
There will very often be affiliations in common between the different authors. Like for affiliation normalisation, you should do it only once per affiliation instead of doing it again for every author.